### PR TITLE
Switch node network stack to use native fetch

### PR DESCRIPTION
**Pull request by pkukielka**
## Changes

That PR started as attempt to fix pollyjs recording process. But to fix it, I had to first deal with all of the issues mentioned bellow. I could probably create several PRs with more isolated changes, but not with a properly working tests.
Thus, I decided that doing big PR with all those changes, BUT also with working tests is a safer choice.

Our network stack was really messy. At the same time we were using:

* native fetch
* isomorphic-fetch
* http.request

This PR unifies the stack, making native fetch a default for all node clients (web is still using isomorphic-fetch).
To do that, completions streaming was rewritten from `http.request` to `fetch` streams.

Unifying the stack makes it possible to handle things properly on the pollyjs end.
I switched adapter we were using from `@pollyjs/adapter-node-http` to `@pollyjs/adapter-fetch` - `adapter-node-http` mocks `http.request` instead of fetch and is not suitable for our needs anymore.

With all of the above I still had to fix all the failing tests. I noticed several problems we had with auth propagation in `agent.ts`. Issues are mainly caused by incorrect usage of `currentAuthStatus`. We are propagating auth status in one place, and then call `currentAuthStatus` in other a moment later. Hidden assumption is that it was already updated, but there is no quarantine of that. I have seen this problem ([before](https://github.com/sourcegraph/cody/pull/6159/commits/cb92085b5c7625d7e331a44db4b2edbe040938dc))
and now I fixed in couple of other places.
I think `currentAuthStatus` is too easy to use incorrectly, we should consider deprecating it and use `authStatus.map` instead (or `currentAuthStatusOrNotReadyYet` if we really have to).

After all of the changes above - re-recording of pollyjs requests was possible and tests are passing
 
## Test plan

1. Automatic test coverage through fixed pollyjs tests
5. Full QA

### CODE REVIEW

File: vscode/src/completions/nodeClient.ts [hunk 1]
```diff
@@ -37,243 +31,101 @@ export class SourcegraphNodeCompletionsClient extends SourcegraphCompletionsClie
...
-
-            onAbort(signal, () => request.destroy())
+            } catch (error) {
+                const errorObject = error instanceof Error ? error : new Error(`${error}`)
+                log?.onError(errorObject.message, error)
+                recordErrorToSpan(span, errorObject)
+                cb.onError(errorObject)
+            }
         })
     }
```
Comment (pkukielka) [inline vscode/src/completions/nodeClient.ts:131]: I'm wondering if we even need two implementations now.
If streaming is not supported shouldn't streaming API just return one chunk containing whole body?
What would be easiest way to test this in practice -  e.g on which instance and with what model?

File: vscode/src/services/AuthProvider.ts [hunk 2]
```diff
@@ -180,7 +180,7 @@ class AuthProvider implements vscode.Disposable {
         config: ResolvedConfigurationCredentialsOnly | AuthCredentials,
         mode: 'store-if-valid' | 'always-store',
         signal?: AbortSignal
-    ): Promise<{ isStored: boolean; authStatus: AuthStatus }> {
```
Comment (pkukielka) [inline vscode/src/services/AuthProvider.ts:183]: `isStored` was never used and was only complicating reading `authStatus`.

File: vscode/src/completions/nodeClient.ts [hunk 3]
```diff
@@ -13,16 +11,12 @@ import {
     NetworkError,
     RateLimitError,
     SourcegraphCompletionsClient,
-    addClientInfoParams,
     addCodyClientIdentificationHeaders,
     currentResolvedConfig,
+    fetch,
```
Comment (pkukielka) [inline vscode/src/completions/nodeClient.ts:16]: Interestingly for non-streamed completions we were using native fetch already... but that means we were using two different fetch implementations at the same time.

File: agent/recordings/auth_2503977039/recording.har.yaml [hunk 4]
```diff
@@ -12,55 +12,62 @@ log:
...
+            {
+                "codyEnabled": true,
+                "chatEnabled": true,
+                "autoCompleteEnabled": true,
+                "customCommandsEnabled": true,
+                "attributionEnabled": false,
+                "smartContextWindowEnabled": true,
+                "modelsAPIEnabled": false,
+                "latestSupportedCompletionsStreamAPIVersion": 5
+            }
```
Comment (mkondratek) [inline agent/recordings/auth_2503977039/recording.har.yaml:44]: It looks like the recording structure is simpler now and more readable (as the text is no longer encoded) - nice 🚀

File: package.json [hunk 5]
```diff
@@ -31,7 +31,7 @@
     "generate-agent-kotlin-bindings": "./agent/scripts/generate-agent-kotlin-bindings.sh",
     "update-agent-recordings": "pnpm build && CODY_KEEP_UNUSED_RECORDINGS=false CODY_RECORD_IF_MISSING=true vitest agent/src",
     "update-agent-recordings-windows": "PowerShell -ExecutionPolicy Bypass -Command \"pnpm build; if ($?) { $Env:CODY_KEEP_UNUSED_RECORDINGS='false'; $Env:CODY_RECORD_IF_MISSING='true'; vitest agent/src }\"",
-    "update-rewrite-recordings": "rm -rf recordings && CODY_RECORD_IF_MISSING=true vitest vscode/src/local-context/rewrite-keyword-query.test.ts",
+    "update-rewrite-recordings": "rm -rf recordings && CODY_RECORD_IF_MISSING=true CODY_RECORDING_MODE=record vitest vscode/src/local-context/rewrite-keyword-query.test.ts",
```
Comment (mkondratek) [inline package.json:34]: 😮

**Review by mkondratek — COMMENTED** [review]
```
> pnpm run -s build:root && pnpm run -s build:webviews && pnpm run -s build:agent

node:internal/modules/cjs/loader:1251
  throw err;
  ^

Error: Cannot find module '/Users/mkondratek/IdeaProjects/cody/node_modules/.pnpm/vite@5.4.8_@types+node@20.12.7/node_modules/vite/bin/vite.js'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1248:15)
    at Module._load (node:internal/modules/cjs/loader:1074:27)
    at TracingChannel.traceSync (node:diagnostics_channel:315:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:217:24)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:166:5)
    at node:internal/main/run_main_module:30:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}

Node.js v22.6.0

FAILURE: Build failed with an exception.

```

I tried it with the JB client. Build fails with ☝️

**Comment by pkukielka** [comment]
> ```
> > pnpm run -s build:root && pnpm run -s build:webviews && pnpm run -s build:agent
> 
> node:internal/modules/cjs/loader:1251
>   throw err;
>   ^
> 
> Error: Cannot find module '/Users/mkondratek/IdeaProjects/cody/node_modules/.pnpm/vite@5.4.8_@types+node@20.12.7/node_modules/vite/bin/vite.js'
>     at Module._resolveFilename (node:internal/modules/cjs/loader:1248:15)
>     at Module._load (node:internal/modules/cjs/loader:1074:27)
>     at TracingChannel.traceSync (node:diagnostics_channel:315:14)
>     at wrapModuleLoad (node:internal/modules/cjs/loader:217:24)
>     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:166:5)
>     at node:internal/main/run_main_module:30:49 {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: []
> }
> 
> Node.js v22.6.0
> 
> FAILURE: Build failed with an exception.
> ```
> 
> I tried it with the JB client. Build fails with ☝️

Try running `pnpm install` in the cody dir to refresh the dependencies.

**Comment by mkondratek** [comment]
I removed my node_modules. I helped. Then I needed to run it twice b/c the first time showed some UI issues. Now it looks good and works.

File: agent/scripts/export-cody-http-recording-tokens.sh [hunk 6]
```diff
@@ -12,10 +12,11 @@
 # Tips to update these secrets:
 #  - Use 1Password to find account passwords
 #  - Use no expiration dates when creating access tokens
-#  - You need to update the REDACTED_ access token in
-#    agent/src/testing-credentials.ts.  There's no automatic way to do this for now,
-#    you need to re-record with the new token, manually copy the updated REDACTED_
-#    token from the recording file and paste it into agent/src/testing-tokens.ts.
+#  - You need to update the REDACTED_ access token in agent/src/testing-credentials.ts.
+#     First you  you need to re-record with the new token.
```
Comment (dominiccooney) [inline agent/scripts/export-cody-http-recording-tokens.sh:16]:
```suggestion
#    First you need to re-record with the new token.
```

File: agent/src/__snapshots__/custom-commands.test.ts.snap [hunk 7]
```diff
@@ -1,25 +1,20 @@
...
-These are the raw filenames without any additional formatting, as requested."
+• src/example1.ts
+• src/example2.ts
+• src/example3.ts"
 `;
 
 exports[`Custom Commands > commands/custom, edit command, edit mode 1`] = `
-"/* SELECTION_START */
-export interface Animal {
+"export interface Animal {
```
[the remark this answers is not in the document]
Comment (dominiccooney) [reply agent/src/__snapshots__/custom-commands.test.ts.snap:12]: @mkondratek 

No, these are recorded test expectations. They depend on the LLM output; when you update recordings, you can get different update which updates these expectations.

Reviewing this is hard but if the contents are approximately equal... not echoing error messages or completely unrelated content... then they're OK.

File: lib/shared/src/chat/sse-iterator.ts [hunk 8]
```diff
@@ -14,32 +14,40 @@ export async function* createSSEIterator(
...
 
-        let index: number
-        // biome-ignore lint/suspicious/noAssignInExpressions: useful
-        while ((index = buffer.indexOf(SSE_TERMINATOR)) >= 0) {
-            const message = buffer.slice(0, index)
-            buffer = buffer.slice(index + SSE_TERMINATOR.length)
-            messages.push(parseSSEEvent(message))
-        }
+            const messages: SSEMessage[] = []
+            buffer += new TextDecoder().decode(value)
```
Comment (dominiccooney) [inline lib/shared/src/chat/sse-iterator.ts:25]: This is not right: If the message is split within a multibyte character, then decoding text at that boundary will corrupt characters. (The existing code with its `toString` is probably also wrong...)

Instead, either keep the content as bytes until you find the SSE_TERMINATOR and then decode the message text; or wrap the ReadableStream up into a TextDecoderStream.

File: vscode/src/completions/nodeClient.ts (hunk 1 above)
Comment (dominiccooney) [reply vscode/src/completions/nodeClient.ts:133]: These are good questions but short answer is: I don't know. @valerybugakov, @abeatrix do you know?

**Review by dominiccooney — APPROVED** [review]
Approving to unblock.

Please fix the character decoding to not do so on arbitrary byte boundaries noted inline. You can probably test this by asking Cody to generate lots of Japanese or Chinese and noting if you get mojibake like Unicode boxes. Best done on a network that has a small transfer unit size.

Thanks for working on this--being able to make recordings again, and making auth status less flaky in tests--is a huge relief.

File: lib/shared/src/chat/sse-iterator.ts (hunk 8 above)
Comment (pkukielka) [reply lib/shared/src/chat/sse-iterator.ts]: Thank you a lot!
That is very good point, I added fix with the test for this scenario.

**Pull request closed without merging**